Add option to generate arbitrary rectangular (prism) regions.

RyanWollaeger opened the pull request:
### Background

* It can be useful to delineate regions for an X3D mesh, where each region can have different properties ascribed to it.

### Purpose of Pull Request

* Add ability to specify any number of rectangular (prism) regions with arbitrary matids (called reg_ids on command line).
* For the changeset here, the `matid` field in the main X3D file is consistent with the region file cell partitioning (but the region file enumeration - Reg1, Reg2, etc - is not equal to the matid value per region).

### Description of changes

+ Add optional list inputs for region id and region bounds:
  * `--reg_ids` is a list of matids per region,
  * `--reg_bnd_per_dim` is a serialized list of boundaries dimensions, similar to `--bnd_per_dim` but for each value in `--reg_ids` (in the same order). 
+ Sanity check region list sizes if they have been specified.
+ Calculate a basic cell center and check if it is in the region.
+ Use region ids for matid field and create region file output.
+ Default to one universal region (matid=0) if regions unspecified.
+ Add example to README.md.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash3/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/cdash3/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation

RyanWollaeger committed:
Add option to generate arbitrary rectangular (prism) regions.

+ Add optional list inputs for region id and region bounds.
+ Sanity check region list sizes if they have been specified.
+ Calculate a basic cell center and check if it is in the region.
+ Use region ids for matid field and create region file output.
+ Default to one universal region (matid=0) if regions unspecified.
+ Add example to README.md.

RyanWollaeger commented on src/mesh/python/x3d_generator.py:
```diff
@@ -51,6 +59,65 @@
+    nmats = len(reg_ids)
+    for cell in range(mesh.num_cells):
+        # -- initialize container for cell average coordinate
+        coord_av = np.array([0.0, 0.0, 0.0])
```
This cell-average coordinate computation should probably be a method for each mesh (for Voronoi it would just be the original sampled points used to generate the mesh).  However, this would be a fairly minor change (and just a refactor, not a significant feature change).

KineticTheory reviewed (approved):
LGTM - but I just did a cursory review.  I'll leave it to @brryan to check that the update works as advertised.

brryan commented on src/mesh/python/x3d_generator.py:
```diff
@@ -51,6 +59,65 @@
+            # -- add face average to cell average
+            coord_av += coord_face_av
+        # -- finish cell average by dividing by number of faces per cell
+        coord_av /= mesh.num_faces_per_cell[cell]
```
Will this work for all concave cells?

RyanWollaeger replied:
Good question - there are sort of a few answers to this.  The coordinate average here:
1.) can be outside of cell if the cell is concave,
2.) serves as an indicator for whether the cell is included in a particular region,
3.) but the region is merely an overly (e.g painting a label), so doesn't affect the mesh structure,
4.) cells can be included in a region without being entirely inside it, so there is error on the order of the cell resolution.

RyanWollaeger replied:
Putting (1)-(4) together, if a cell is extremly concave, it is possible (pretty sure at least) most of it's volume will be out of the region, while the `coord_av` will be in the region.  This is why I had commented that this averaging should be a method in the `mesh_type` mesh classes (but it didn't seem urgent to do so).  Should I go ahead and make this change here (it is not too difficult)?

brryan replied:
> This is why I had commented that this averaging should be a method in the mesh_type mesh classes

Ah OK I misunderstood that comment above. 

I think this is fine as is for now (especially since, like you say, this is really a spatial error of a type that is already inherent in the mesh scheme), but what do you think about adding a comment somewhere here e.g. `TODO(RW) make coordinate averaging a separate method for each mesh_type to account for concave cells`? (That's 100 characters so you might have a better comment in mind)

RyanWollaeger replied:
Just added a TODO comment for this.

brryan reviewed (commented):
Just one question on whether we are specializing to more-or-less convex polygons for cells here, otherwise LGTM

brryan commented on src/mesh/python/x3d_generator.py:
```diff
@@ -51,6 +59,65 @@
+    reg_ids = args.reg_ids
+    reg_bnd_per_dim = args.reg_bnd_per_dim
+    reg_ids = [0] + reg_ids
+    reg_bnd_per_dim = args.bnd_per_dim + reg_bnd_per_dim
```
Is it worth checking here that a universal region wasn't passed in through the command line args by an unsuspecting user?

RyanWollaeger replied:
Good idea! Will add a check for this...

RyanWollaeger replied:
This may be okay, since the universal region does not get added if all its cells are covered by 1 or more other regions.  However, removing cells from the universal region (originally at line 111) that overlap with other regions was only being done for the universal region, and not for specified regions, which could cause overlap.  To avoid this, in the recent commit line 111 has been generalized to subtract cells from regions overlapping with regions specified later in the user input list (this has been tested on the 2-region example that was added to the README).

RyanWollaeger committed:
Incorporate suggestions from Ben Ryan.

+ Ensure all regions do not overlap in cells.
+ Add TODO for making cell coordinate averaging a mesh_type method.

RyanWollaeger commented:
It appears one coverage check is not passing.  Going to try rebasing this PR against develop to see if it is avoidable.

KineticTheory commented:
> 
> 
> It appears one coverage check is not passing. Going to try rebasing this PR against develop to see if it is avoidable.

Don't bother with the rebase. The coverage is still good. If this is ready, let's just merge it.

RyanWollaeger commented:
> > It appears one coverage check is not passing. Going to try rebasing this PR against develop to see if it is avoidable.
> 
> Don't bother with the rebase. The coverage is still good. If this is ready, let's just merge it.

Sorry - missed your comment and just rebased (doesn't hurt I suppose).

brryan commented:
LGTM
